llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -701,6 +701,42 @@ Value *InstCombinerImpl::simplifyRangeCheck(ICmpInst *Cmp0, ICmpInst *Cmp1,
   return Builder.CreateICmp(NewPred, Input, RangeEnd);
 }
 
+// (or (icmp eq X, 0), (icmp eq X, Pow2OrZero))
+//      -> (icmp eq (and X, Pow2OrZero), X)
+// (and (icmp ne X, 0), (icmp ne X, Pow2OrZero))
+//      -> (icmp ne (and X, Pow2OrZero), X)
+static Value *
+foldAndOrOfICmpsWithPow2AndWithZero(InstCombiner::BuilderTy &Builder,
+                                    ICmpInst *LHS, ICmpInst *RHS, bool IsAnd,
+                                    const SimplifyQuery &Q) {
+  CmpInst::Predicate Pred = IsAnd ? CmpInst::ICMP_NE : CmpInst::ICMP_EQ;
+  // Make sure we have right compares for our op.
+  if (LHS->getPredicate() != Pred || RHS->getPredicate() != Pred)
+    return nullptr;
+
+  // Make it so we can match LHS against the (icmp eq/ne X, 0) just for
+  // simplicity.
+  if (match(RHS->getOperand(1), m_Zero()))
+    std::swap(LHS, RHS);
+
+  Value *Pow2, *Op;
+  // Match the desired pattern:
+  // LHS: (icmp eq/ne X, 0)
+  // RHS: (icmp eq/ne X, Pow2OrZero)
+  // We could potentially be more sophisticated than requiring LHS/RHS be
+  // one-use. We don't create additional instructions if only one of them is
+  // one-use. So cases where one is one-use and the other is two-use might be
+  // profitable.
+  if (!match(LHS, m_OneUse(m_ICmp(Pred, m_Value(Op), m_Zero()))) ||
+      !match(RHS, m_OneUse(m_c_ICmp(Pred, m_Specific(Op), m_Value(Pow2)))) ||
+      !isKnownToBeAPowerOfTwo(Pow2, Q.DL, /*OrZero=*/true, /*Depth=*/0, Q.AC,
```

Comment:
We shouldn't fold `and (icmp ne X, 0), (icmp ne X, 1) -> icmp ne (and X, 1), X`.
See https://github.com/dtcxzyw/llvm-opt-benchmark/pull/667/files#r1632021232.

It will break the `x != 0 && x != 1 && x != 2` pattern:
https://github.com/php/php-src/blob/d1f14a46095571b1b6363fa7f83c6a681d1d02de/main/main.c#L453-L456

Reply:
I guess it doesn't matter either way, it always turns into `x u> 1`: https://godbolt.org/z/4vE4nMKTb
although we should teach whichever pattern is looking at consecutative `ne` to handle that

Reply:
Ill forbid `1`, also maybe should skip if `Op` is an `and` expression as well? That seemed to be the common element in all other regressions.

Reply:
Yeah, but there is a phase ordering problem blocking this fold:
https://github.com/dtcxzyw/llvm-opt-benchmark/pull/667/files#r1632021232

What do you think of the following fold?
https://alive2.llvm.org/ce/z/tmkPvm
```
define i1 @src(i16 %x) {
 %35 = and i16 %x, 255
 call void @use(i16 %35)
  %36 = and i16 %x, 1
  %or.cond = icmp ne i16 %36, %35
  ret i1 %or.cond
}

define i1 @tgt(i16 %x) {
  %35 = and i16 %x, 255
  call void @use(i16 %35)
  %cmp = icmp ugt i16 %35, 1
  ret i1 %cmp
}

declare void @use(i16)
```

Reply:
Fixed `1` cae + tests

Reply:
Emm, this regression still exists :( We may need more folds to fix this.